### ISSUE TITLE
Fingerprint differs for a file although src and target are equal

### ISSUE THREAD

**Issue by sebwarnke** [bcpierce00/unison#739]
Hi,

OS: macOS 12.4
CPU: Apple Silicon M1

My profile:
```
root = /Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney
root = /Volumes/admin/Backup/MoneyMoney
ignore = Name {Icon^M}
ignore = Name {Icon}
ignore = Name {Icon?}
ignore = Name {.DS_Store}
ignore = Name {MoneyMoney-Backup.sqlite}
ignore = Name {MoneyMoney.sqlite-journal}
ignore = Path {Statements}
perms = 0
```

`/Volumes/admin/Backup/MoneyMoney` is a smb mount from close proximity. No VPN, no latency, nothing.

In Unison 2.52.1 copying a certain file using the profile above fails with the following error;

```
Fingerprints differ: (527d201d8163645c0d1faa506cd069f9,c4586ab37f6b15d1714aab14e960a161) vs (527d201d8163645c0d1faa506cd069f9,)
```
from the below debug log.
```
Unison 2.52.1 (ocaml 4.14.0) started propagating changes at 23:11:00.69 on 25 May 2022
<>$<>Starting new major GC cycle
[BGN] Updating file Database/MoneyMoney.sqlite from /Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney to /Volumes/admin/Backup/MoneyMoney
[files] copy /Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney Database/MoneyMoney.sqlite ---> /Volumes/admin/Backup/MoneyMoney Database/MoneyMoney.sqlite
[pred] follow 'Database/MoneyMoney.sqlite' = false
[fspath] Os.findWorkingDir(/Volumes/admin/Backup/MoneyMoney,Database/MoneyMoney.sqlite) = (/Volumes/admin/Backup/MoneyMoney/Database,MoneyMoney.sqlite)
[update] updateArchive /Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney Database/MoneyMoney.sqlite
[pred] ignore 'Database/MoneyMoney.sqlite' = false
[files] copyRec Database/MoneyMoney.sqlite --> .unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp  (really to MoneyMoney.sqlite)
[abort] Checking line 0
[copy] copyRegFile(/Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney,Database/MoneyMoney.sqlite) -> (/Volumes/admin/Backup/MoneyMoney,MoneyMoney.sqlite,/Volumes/admin/Backup/MoneyMoney/Database,.unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp,modified on 2022-05-25 at 23:10:26  size 15202304  unknown permissions)
[copy] Copy.localFile /Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney / Database/MoneyMoney.sqlite to /Volumes/admin/Backup/MoneyMoney/Database / .unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp
[osx] openRessOut /Volumes/admin/Backup/MoneyMoney/Database/.unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp/..namedfork/rsrc
[osx] Opening AppleDouble file for resource fork
[props] Setting permissions for /Volumes/admin/Backup/MoneyMoney/Database/.unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp to unknown permissions (700/0)
<>![copy] Fingerprints differ: (527d201d8163645c0d1faa506cd069f9,c4586ab37f6b15d1714aab14e960a161) vs (527d201d8163645c0d1faa506cd069f9,)
[copy] TRANSFER FAILED (resource fork) for .unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp (real path: MoneyMoney.sqlite)
[copy] Failed (resource fork): Saving old temp file .unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp
[os] rename /Volumes/admin/Backup/MoneyMoney/.unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp to /Volumes/admin/Backup/MoneyMoney/.unison.MoneyMoney.sqlite-bad.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp
[exn] Converting a Unix error to Transient:
Error in renaming /Volumes/admin/Backup/MoneyMoney/.unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp to /Volumes/admin/Backup/MoneyMoney/.unison.MoneyMoney.sqlite-bad.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp:
No such file or directory [rename(/Volumes/admin/Backup/MoneyMoney/.unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp)]
[abort] Aborting line 0
Failed: The file .unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp was incorrectly transferred  (fingerprint mismatch in resource fork) -- temp file saved as /Volumes/admin/Backup/MoneyMoney/.unison.MoneyMoney.sqlite-bad.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp
Unison 2.52.1 (ocaml 4.14.0) finished propagating changes at 23:11:02.86 on 25 May 2022, 2.170 s
```
**Some observations:**

- The files `MoneyMoney.sqlite` and `.unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp` are entirely equal by `shasum`, `cmp -l` and `diff`.
- The issue only happens to this particular file.
- When roots are on the Mac's SDD both, copying the file works.
- Unison 2.51.2 does not have that issue.

Does anyone have an idea what change between the version might cause this?

**Comment by tleedjarv** [comment]
The issue is related to `MoneyMoney.sqlite` having a resource fork (is that likely?) that fails to be copied to the smb mount.

You can try adding `rsrc = false` to your profile. This will stop syncing resource forks, so don't use it if you know that you want to sync them.

**Comment by tleedjarv** [comment]
When running Unison 2.51.2, if you enable debugging, do you see lines like this in the debug log?
```
[osx] openRessOut /Volumes/admin/Backup/MoneyMoney/Database/.unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp/..namedfork/rsrc
[osx] Opening AppleDouble file for resource fork
```

Additionally, do you see a line containing text `resource fork fingerprint`?

**Comment by gdt** [comment]
I am unclear on whether there is a unison bug here or not.   I can't tell fi "resource fork is not copied" is about SMB being deficient, or if it's simply "target FS is not mac", and what should be happening.  It would be good to close this and have a clean bug description if there is a bug, and to close this and move it to the mailinglist (as the issue tracker is not a support forum) if there is not a bug.

**Comment by tleedjarv** [comment]
To be fair, the report has plenty of quality information. It's great if `rsrc = false` works for @sebwarnke but there is a hint of a regression here:

> * Unison 2.51.2 does not have that issue.

As to your question about what should be happening. It should just work, even if target is not a mac. It apparently works with 2.51.2 but this can be confirmed when @sebwarnke responds to questions about debug log above.

> or if it's simply "target FS is not mac"

That's not the issue. There is code to cover that. Line `[osx] Opening AppleDouble file for resource fork` in the log indicates that this is the "not being a mac" code path taken.

**Comment by sebwarnke** [comment]
> The issue is related to `MoneyMoney.sqlite` having a resource fork (is that likely?) that fails to be copied to the smb mount.
> 
> You can try adding `rsrc = false` to your profile. This will stop syncing resource forks, so don't use it if you know that you want to sync them.

@tleedjarv, thank you! `rsrc = false` resolves the issue in Unison 2.52.1.

> When running Unison 2.51.2, if you enable debugging, do you see lines like this in the debug log?
> 
> ```
> [osx] openRessOut /Volumes/admin/Backup/MoneyMoney/Database/.unison.MoneyMoney.sqlite.8e7e4c8c2e7db1f7e53d4e79c2a84716.unison.tmp/..namedfork/rsrc
> [osx] Opening AppleDouble file for resource fork
> ```

Nope, running Unison 2.51.2 does not create the above log, nor `resource fork fingerprint`;

```
UNISON 2.51.2 (OCAML 4.04.0) started propagating changes at 22:00:06.73 on 26 May 2022
[BGN] Updating file Database/MoneyMoney.sqlite from /Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney to /Volumes/admin/Backup/MoneyMoney
[files] copy /Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney Database/MoneyMoney.sqlite ---> /Volumes/admin/Backup/MoneyMoney Database/MoneyMoney.sqlite
[pred] follow 'Database/MoneyMoney.sqlite' = false
[fspath] Os.findWorkingDir(/Volumes/admin/Backup/MoneyMoney,Database/MoneyMoney.sqlite) = (/Volumes/admin/Backup/MoneyMoney/Database,MoneyMoney.sqlite)
[update] updateArchive /Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney Database/MoneyMoney.sqlite
[pred] ignore 'Database/MoneyMoney.sqlite' = false
[files] copyRec Database/MoneyMoney.sqlite --> .unison.MoneyMoney.sqlite.621d50d3727b7e7ac757c2520b4f9c3f.unison.tmp  (really to MoneyMoney.sqlite)
[abort] Checking line 0
[copy] copyRegFile(/Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney,Database/MoneyMoney.sqlite) -> (/Volumes/admin/Backup/MoneyMoney,MoneyMoney.sqlite,/Volumes/admin/Backup/MoneyMoney/Database,.unison.MoneyMoney.sqlite.621d50d3727b7e7ac757c2520b4f9c3f.unison.tmp,modified on 2022-05-26 at 21:59:47  size 15202304  unknown permissions)
[copy] Copy.localFile /Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney / Database/MoneyMoney.sqlite to /Volumes/admin/Backup/MoneyMoney/Database / .unison.MoneyMoney.sqlite.621d50d3727b7e7ac757c2520b4f9c3f.unison.tmp
[props] Setting permissions for /Volumes/admin/Backup/MoneyMoney/Database/.unison.MoneyMoney.sqlite.621d50d3727b7e7ac757c2520b4f9c3f.unison.tmp to unknown permissions (700/0)
[files] rename(root=/Volumes/admin/Backup/MoneyMoney, localPath=Database/MoneyMoney.sqlite, pathOld=.unison.MoneyMoney.sqlite.621d50d3727b7e7ac757c2520b4f9c3f.unison.tmp, pathNew=MoneyMoney.sqlite)
[update] checkNoUpdates /Volumes/admin/Backup/MoneyMoney Database/MoneyMoney.sqlite
[update] buildUpdateRec: /Volumes/admin/Backup/MoneyMoney/Database/MoneyMoney.sqlite
[update] checkContentsChange: archStamp is inode (33947658) / info.inode (33947658) / times: 1653594902.000000 = 1653594902.000000... true / lengths: 15202304 - 15202304
[update]   archive digest = (3344855a44a77b7f06197f53ec5c2476,)   current digest = (3344855a44a77b7f06197f53ec5c2476,)
[files] Renaming .unison.MoneyMoney.sqlite.621d50d3727b7e7ac757c2520b4f9c3f.unison.tmp to MoneyMoney.sqlite in /Volumes/admin/Backup/MoneyMoney/Database; root is /Volumes/admin/Backup/MoneyMoney
[files] rename: moveFirst=false
[stasher] backup: /Volumes/admin/Backup/MoneyMoney / Database/MoneyMoney.sqlite
[pred] follow 'Database/MoneyMoney.sqlite' = false
[fspath] Os.findWorkingDir(/Volumes/admin/Backup/MoneyMoney,Database/MoneyMoney.sqlite) = (/Volumes/admin/Backup/MoneyMoney/Database,MoneyMoney.sqlite)
[pred] backup 'Database/MoneyMoney.sqlite' = false
[stasher] Path /Volumes/admin/Backup/MoneyMoney / Database/MoneyMoney.sqlite does not need to be backed up
[os] rename /Volumes/admin/Backup/MoneyMoney/Database/.unison.MoneyMoney.sqlite.621d50d3727b7e7ac757c2520b4f9c3f.unison.tmp to /Volumes/admin/Backup/MoneyMoney/Database/MoneyMoney.sqlite
[files] Contents of /Volumes/admin/Backup/MoneyMoney/Database/MoneyMoney.sqlite after renaming = bdefb0814d330184612fb4a3615129f3
[pred] backupcurr 'Database/MoneyMoney.sqlite' = false
[update] replaceArchiveLocal /Volumes/admin/Backup/MoneyMoney Database/MoneyMoney.sqlite
[update] Setting archive for //AwesomeParkrs.fritz.box//Volumes/admin/Backup/MoneyMoney
[update] updateArchive /Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney Database/MoneyMoney.sqlite
[pred] ignore 'Database/MoneyMoney.sqlite' = false
[update] replaceArchiveLocal /Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney Database/MoneyMoney.sqlite
[update] Setting archive for //AwesomeParkrs.fritz.box//Users/seb/Library/Containers/com.moneymoney-app.retail/Data/Library/Application Support/MoneyMoney
[pred] backupcurr 'Database/MoneyMoney.sqlite' = false
[END] Updating file Database/MoneyMoney.sqlite
UNISON 2.51.2 (OCAML 4.04.0) finished propagating changes at 22:00:07.24 on 26 May 2022


[update] Updating archives
[update] Saving archive in /Users/seb/Library/Application Support/Unison/sc23631bbbb90830fa4842421e65db4bfa
[update] Saving archive in /Users/seb/Library/Application Support/Unison/sc621d50d3727b7e7ac757c2520b4f9c3f
[update] Copying archive /Users/seb/Library/Application Support/Unison/tm23631bbbb90830fa4842421e65db4bfa to /Users/seb/Library/Application Support/Unison/ar23631bbbb90830fa4842421e65db4bfa
[update] Copying archive /Users/seb/Library/Application Support/Unison/tm621d50d3727b7e7ac757c2520b4f9c3f to /Users/seb/Library/Application Support/Unison/ar621d50d3727b7e7ac757c2520b4f9c3f
[update] Removing archive /Users/seb/Library/Application Support/Unison/tm23631bbbb90830fa4842421e65db4bfa
[update] Removing archive /Users/seb/Library/Application Support/Unison/tm621d50d3727b7e7ac757c2520b4f9c3f
```

**Comment by sebwarnke** [comment]
Thanks for your help again, @tleedjarv.
